tests/cases/compiler/restParamUsingMappedTypeOverUnionConstraint.ts

```diff
@@ -0,0 +1,11 @@
+// @noEmit: true
+// @strict: true
+
+// repro 29919#issuecomment-470948453
+
+type Mapped<T> = { [P in keyof T]: T[P] extends string ? [number] : [string] }
+
+declare function test<T extends [number] | [string]>(
```

Comment:
This repro is a lot easier to understand for me because it has fewer moving parts and fewer repeated identifier, so it's easier to tell what dependencies exist between elements:

```ts
declare function test<T extends [number] | []>(
    fn: (...args: { [_ in keyof T]: [boolean] | [string] }) => void
): number
```

Reply:
I agree with the fact that the proposed version is more succinct but at the same time, I don't feel like the added test case is overly verbose. For me personally, this is a **little** bit harder to read because it's less "usual". The inlined mapped type also doesn't make it completely obvious to me if this mapped type is still homomorphic or not (although that's just my lack of expertise).

Overall, I don't mind making a change but perhaps it would be good to have some guidelines in the `CONTRIBUTING.md` for how one should write tests etc. For an external contributor like me, it's not that obvious what kind of tests are preferred by the team.

Please let me know if I should adjust the test, add a new one or leave it as is.